contracts/treasury/priceOracle/KyberDmmChainLinkPriceOracle.sol

```diff
@@ -0,0 +1,382 @@
+        (minAmountOuts[0], minAmountOuts[1]) = (amounts[2], amounts[3]);
+      } else {
+        require(tokens[1] == tokenOuts[0], 'invalid token out 0');
+        (minAmountOuts[0], minAmountOuts[1]) = (amounts[3], amounts[2]);
+      }
+      return _applyPremiumFor(liquidator, minAmountOuts);
+    }
+
+    require(tokenOuts.length == 1, 'invalid number token out');
+    require(isWhitelistedToken(address(tokenOuts[0])), 'token out must be whitelisted');
+
+    // Liquidate list of tokens to a single dest token
+    if (hintType == OracleHintType.LIQUIDATE_TOKENS) {
+      for(uint256 i = 0; i < tokenIns.length; i++) {
+        require(!isWhitelistedToken(address(tokenIns[i])), 'token in can not be a whitelisted token');
+        uint256 rate = conversionRate(address(tokenIns[i]), address(tokenOuts[0]), amountIns[i]);
+        require(rate > 0, 'invalid conversion rate');
+        minAmountOuts[0] = minAmountOuts[0].add(
+          calcDestAmount(tokenIns[i], tokenOuts[0], amountIns[i], rate)
+        );
+      }
+      return _applyPremiumFor(liquidator, minAmountOuts);
+    }
+
+    // Liquidate list of LP tokens to a single dest token
+    for(uint256 i = 0; i < tokenIns.length; i++) {
+      (IERC20Ext[2] memory tokens, uint256[4] memory amounts) = getExpectedTokensFromLp(
+        address(tokenIns[i]), amountIns[i]
+      );
+      // calc equivalent (tokens[0], amounts[2]) -> tokenOuts[0]
+      uint256 rate = conversionRate(address(tokens[0]), address(tokenOuts[0]), amounts[2]);
+      require(rate > 0, 'invalid conversion rate 0');
+      minAmountOuts[0] = minAmountOuts[0].add(
+        calcDestAmount(tokens[0], tokenOuts[0], amounts[2], rate)
+      );
+      // calc equivalent (tokens[1], amounts[3]) -> tokenOuts[0]
+      rate = conversionRate(address(tokens[1]), address(tokenOuts[0]), amounts[3]);
+      require(rate > 0, 'invalid conversion rate 1');
+      minAmountOuts[0] = minAmountOuts[0].add(
+        calcDestAmount(tokens[1], tokenOuts[0], amounts[3], rate)
+      );
+    }
+    minAmountOuts = _applyPremiumFor(liquidator, minAmountOuts);
+  }
+
+  // Whitelisted tokens
+  function getWhitelistedTokensLength() external view returns (uint256) {
+    return _whitelistedTokens.length();
+  }
+
+  function getWhitelistedTokenAt(uint256 index) external view returns (address) {
+    return _whitelistedTokens.at(index);
+  }
+
+  function getAllWhitelistedTokens()
+    external view returns (address[] memory tokens)
+  {
+    uint256 length = _whitelistedTokens.length();
+    tokens = new address[](length);
+    for(uint256 i = 0; i < length; i++) {
+      tokens[i] = _whitelistedTokens.at(i);
+    }
+  }
+
+  /**
+   * @dev Return expect amounts given pool and number of lp tokens
+   *  TODO: should have a solution for token with fees, or just another contract to support
+   * @return tokens [token0, token1]
+   * @return amounts [amount0, amount1, expectedAmount0, expectedAmount1s]
+   */
+  function getExpectedTokensFromLp(
```

Comment:
seems like the returned reserve amounts (indexes 0 and 1) are unused, can consider not returning them.

Reply:
We don't verify that token is dmm Lp token pool?

Reply:
There is no way to verify, and we don't need actually

Reply:
https://github.com/dynamic-amm/smart-contracts/blob/master/contracts/periphery/DMMRouter02.sol#L656

Reply:
Actually, we can remove lines 195-202 by this check